### PR TITLE
Refactors MoreLikeThisQueryBuilder and Parser

### PR DESCRIPTION
**Pull request by alexksikes**
Relates to #10217

This PR is against the query-refactoring branch.

### CODE REVIEW

**Comment by s1monw** [comment]
@alexksikes can you rebase this to latest - there are quite some changes here that are now gone

**Comment by alexksikes** [comment]
@s1monw You can take a look now. Thank you.

File: core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java [hunk 1]
```diff
@@ -147,7 +204,8 @@ public Item doc(BytesReference doc) {
          * Sets to a given artificial document, that is a document that is not present in the index.
          */
         public Item doc(XContentBuilder doc) {
-            return this.doc(doc.bytes());
+            this.doc = doc != null ? doc.bytes() : null;
```
Comment (colings86) [inline core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java]: I don't think we should allow null here I think this should throw an exception if null is passed in

Comment (colings86) [inline core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java]: In fact, given that we require either doc or an index document id, this should really be a constructor and this setter should be removed. That way the user can only specify doc OR id and must specify one of them. We should add null checks in the constructors and throw IllegalArgumentException if nulls are passed in

Comment (colings86) [inline core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java]: Also, given that type and index are required if id is specified we should remove the setters for type, id and index and remove the zero-arg constructor for Item so the user is forced to correct initialise the object with either the index, type and id, or the index, type and doc

File: core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java [hunk 2]
```diff
@@ -326,6 +384,48 @@ public final String toString() {
...
+            if (id != null && doc != null) {
+                validationException = validationException == null ? new QueryValidationException() : validationException;
+                validationException.addValidationError("either 'id' or 'doc' can be specified, but not both!");
+            }
+            return validationException;
+        }
+
+        @Override
+        public Item readFrom(StreamInput in) throws IOException {
+            Item item = new Item(in.readOptionalString(), in.readOptionalString(), in.readOptionalString());
```
Comment (colings86) [inline core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java]: This will need to change to read a boolean to tell it whether it needs to read a doc or Id so it can use the right constructor and doesn't trip up passing nulls into it

File: core/src/test/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilderTests.java [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+            queryBuilder.minimumShouldMatch(randomMinimumShouldMatch());
+        }
+        if (randomBoolean()) {
+            queryBuilder.boostTerms(randomFloat() * 10);
+        }
+        if (randomBoolean()) {
+            queryBuilder.include(true);
+        }
+        if (randomBoolean()) {
+            queryBuilder.failOnUnsupportedField(false);
```
Comment (colings86) [inline core/src/test/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilderTests.java]: Why can't this and the one above be randomised?

File: core/src/test/java/org/elasticsearch/index/query/SimpleIndexQueryParserTests.java [hunk 4]
```diff
@@ -99,43 +71,11 @@
...
-import static org.hamcrest.Matchers.closeTo;
-import static org.hamcrest.Matchers.containsString;
-import static org.hamcrest.Matchers.equalTo;
-import static org.hamcrest.Matchers.instanceOf;
-import static org.hamcrest.Matchers.is;
-import static org.hamcrest.Matchers.not;
-import static org.hamcrest.Matchers.notNullValue;
-import static org.hamcrest.Matchers.nullValue;
-import static org.hamcrest.Matchers.sameInstance;
+import static org.hamcrest.Matchers.*;
```
Comment (colings86) [inline core/src/test/java/org/elasticsearch/index/query/SimpleIndexQueryParserTests.java]: Can you revert these changes as it's not a functional and will created conflicts to resolve when master is merged

**Comment by colings86** [comment]
@alexksikes I left some comments. I also think someone else will need to quickly look before its merged as I have no experience of the MLT query

File: core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java (hunk 1 above)
Comment (alexksikes) [inline core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java]: There is a constructor for doc. However, we do have this method for bwc, so maybe we should keep it?

Concerning null index, or type, this is so that these could be initialized by the search context, and we only have this context when the query is actually being created.

File: core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java (hunk 2 above)
Comment (alexksikes) [inline core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java]: Right now if you use a non null doc, then the id is ignored. If we remove the doc setter then yes we would have to do as you propose.

File: core/src/test/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilderTests.java (hunk 3 above)
Comment (alexksikes) [inline core/src/test/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilderTests.java]: By default it is true, so then here we set it to false randomly.

File: core/src/test/java/org/elasticsearch/index/query/SimpleIndexQueryParserTests.java (hunk 4 above)
Comment (alexksikes) [inline core/src/test/java/org/elasticsearch/index/query/SimpleIndexQueryParserTests.java]: Yes we should do this upstream or revert this change. This is just running a optimize import on this class, which has way too many individual imports.

**Comment by alexksikes** [comment]
@colings86 Thanks for the review. I'll address the comments on the next round of review.

File: core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java (hunk 1 above)
Comment (colings86) [inline core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java]: We don't need to keep the constructors/methods for bwc. For the Java API we are able to break backwards compatibility here and we should if it makes the use of the API less trappy (as it does here).

File: core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java (hunk 2 above)
Comment (colings86) [inline core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java]: We should do this

File: core/src/test/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilderTests.java (hunk 3 above)
Comment (colings86) [inline core/src/test/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilderTests.java]: We should set random boolean here too. We should be testing both the default (unset) and using the method will all allowed values (i.e. true and false)

File: core/src/test/java/org/elasticsearch/index/query/SimpleIndexQueryParserTests.java (hunk 4 above)
Comment (colings86) [inline core/src/test/java/org/elasticsearch/index/query/SimpleIndexQueryParserTests.java]: We shouldn't change classes just to clean up imports IMO. Unless there is an actual functional change to the class this just causes messy merges. Also, I don;t see why the number of individual imports makes any difference. Imports are hidden and automated by most IDEs and optimised out by the compile at compile time so have no affect on runtime (IIRC)

File: core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java [hunk 5]
```diff
@@ -316,6 +316,18 @@ public void writeStringArrayNullable(@Nullable String[] array) throws IOExceptio
         }
     }
 
+    /**
+     * Writes a string array, for nullable string, writes it as 0 (empty string).
+     */
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java]: Comment seems wrong, as far as I can see the difference to `writeStringArrayNullable` is that it writes boolean instead of int in case of null argument.

Comment (colings86) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java]: why is this method needed if we have `writeStringArrayNullable`?

Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java]: That one doesn't seem to differentiate between null and zero-length array, which we need if we e.g. test for equality of queries after serialization. At least that's what I think, @ale might have had other reasons.

Comment (alexksikes) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java]: This exactly the reason. I first tried `writeStringArrayNullable`, but then you read back an empty array, not a null array.

File: core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java (hunk 1 above)
Comment (alexksikes) [inline core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java]: OK then if we are allowed to break bwc, then I much prefer to remove this method. That would simplify things from code and user perspective. A definite +1.

File: core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java (hunk 5 above)
Comment (colings86) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java]: I'm not a fan of having both of these methods. The difference between them is too subtle. I think we should instead replace `writeStringArrayNullable` with `writeOptionalStringArray` as the former is trappy and could easily lead to bugs. Serialisation code should always serialise and de-serialise to the same thing.

File: core/src/test/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilderTests.java (hunk 3 above)
Comment (alexksikes) [inline core/src/test/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilderTests.java]: OK sounds good, however we do now have 2/3 of chance to get true, but I understand that in general we should be testing unset.

File: core/src/test/java/org/elasticsearch/index/query/SimpleIndexQueryParserTests.java (hunk 4 above)
Comment (alexksikes) [inline core/src/test/java/org/elasticsearch/index/query/SimpleIndexQueryParserTests.java]: Sure makes sense, I'll change that.

File: core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java (hunk 5 above)
Comment (alexksikes) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java]: +1 should this change be part of this PR? Looks like this is not used anywhere else.

Comment (colings86) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java]: If it's not used anywhere else (or even if its only used in a couple of places) then I would be +1 to doing it in this PR

File: core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java [hunk 6]
```diff
@@ -46,10 +61,50 @@
...
+    // other parameters
+    private boolean failOnUnsupportedField = DEFAULT_FAIL_ON_UNSUPPORTED_FIELDS;
+
+    static final MoreLikeThisQueryBuilder PROTOTYPE = new MoreLikeThisQueryBuilder();
+
     /**
      * A single item to be used for a {@link MoreLikeThisQueryBuilder}.
      */
-    public static final class Item implements ToXContent {
+    public static final class Item implements ToXContent, Writeable<Item> {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java]: Given the amount of code in Item, I wonder if it would be good to move this to its own class. Also, lots of the complexity of this class seems to come from the basic two modes of operation (either as artificial doc or as indexed doc with index/type/id location). Maybe introducing two separate subclasses would help to seperate this. This shouldn't holp up this PR, maybe something for a follow-up, wdyt?

Comment (colings86) [inline core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java]: +1

File: core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryParser.java [hunk 7]
```diff
@@ -78,29 +55,44 @@
         ParseField FAIL_ON_UNSUPPORTED_FIELD = new ParseField("fail_on_unsupported_field");
     }
 
+    public MoreLikeThisQueryParser() {
+
+    }
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryParser.java]: nit: this seems unused

File: core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryParser.java [hunk 8]
```diff
@@ -116,37 +108,29 @@ public Query parse(QueryShardContext context) throws IOException, QueryParsingEx
...
+                    minimumShouldMatch = parser.text();
                 } else if (parseContext.parseFieldMatcher().match(currentFieldName, Field.BOOST_TERMS)) {
-                    float boostFactor = parser.floatValue();
-                    if (boostFactor != 0) {
-                        mltQuery.setBoostTerms(true);
-                        mltQuery.setBoostTermsFactor(boostFactor);
-                    }
-                } else if (parseContext.parseFieldMatcher().match(currentFieldName, Field.MINIMUM_SHOULD_MATCH)) {
-                    mltQuery.setMinimumShouldMatch(parser.text());
-                } else if ("analyzer".equals(currentFieldName)) {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryParser.java]: Just for clarification, were those two clauses duplicates in the present version of the parser?

File: core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java (hunk 6 above)
Comment (alexksikes) [inline core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java]: I agree this is what I first did in this pre-refactoring [PR](https://github.com/elastic/elasticsearch/pull/13372), but had to backtrack to keep bwc. Question: is query refactoring 3.0? How much are we allowed to break?

File: core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java (hunk 5 above)
Comment (alexksikes) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java]: My bad this is actually used in a couple of places. I'll open an issue on this.

File: core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java (hunk 6 above)
Comment (colings86) [inline core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java]: We should break anything we need to to make the Java API clean, easy to use, and less trappy. Given this is only going into 3.0 we should take this opportunity.

Comment (alexksikes) [inline core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java]: I like that, thanks for the clarification.

Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java]: +1, only I don't think we need to hold this PR for this, this is big as it is already. I agree with Colin on cleaning up setters, constructors etc. for Item and get rid of as much ambiguity there as possible, but I think this is easier as a follow up

**Comment by cbuescher** [comment]
@alexksikes I did a round of reviews, given the amount of complexity in constructing the final MLT query I didn't check the internals of the helper methods moved over from the parsers, maybe those can be simplified later but as far as separating the toQuery/fromXContent I think this is good. Maybe someone else should have final look after you adressed the last comments.

File: core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryParser.java (hunk 8 above)
Comment (alexksikes) [inline core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryParser.java]: I noticed that too. There are not duplicates in master but they are in the query refactoring branch. Probably something that happened during merge.

File: core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryParser.java (hunk 7 above)
Comment (alexksikes) [inline core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryParser.java]: Sure I can remove. Note all parser seem to be having this empty unused constructor.

File: core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java (hunk 2 above)
Comment (alexksikes) [inline core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java]: +1

File: core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java (hunk 6 above)
Comment (alexksikes) [inline core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java]: Agreed I'll do the package change at last.

**Comment by alexksikes** [comment]
@cbuescher @colings86 I addressed all the comments. Should I do the package change for the Item class as part of this PR too? Thanks again for the review.

File: core/src/main/java/org/elasticsearch/index/VersionType.java [hunk 9]
```diff
@@ -304,4 +311,18 @@ public static VersionType fromValue(byte value) {
         }
         throw new IllegalArgumentException("No version type match [" + value + "]");
     }
+
+    @Override
+    public VersionType readFrom(StreamInput in) throws IOException {
+        return VersionType.values()[in.readVInt()];
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/VersionType.java]: can we check the value we are reading here? an assertion would be ok

File: core/src/main/java/org/elasticsearch/index/VersionType.java [hunk 10]
```diff
@@ -304,4 +311,18 @@ public static VersionType fromValue(byte value) {
...
+    }
+
+    public static VersionType readVersionTypeFrom(StreamInput in) throws IOException {
+        return PROTOTYPE.readFrom(in);
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeVInt(this.ordinal());
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/VersionType.java]: `this` is unnecessary

File: core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java [hunk 11]
```diff
@@ -46,10 +61,50 @@
...
+    public static final int DEFAULT_MIN_WORD_LENGTH = XMoreLikeThis.DEFAULT_MIN_WORD_LENGTH;
+    public static final int DEFAULT_MAX_WORD_LENGTH = XMoreLikeThis.DEFAULT_MAX_WORD_LENGTH;
+    public static final String DEFAULT_MINIMUM_SHOULD_MATCH = MoreLikeThisQuery.DEFAULT_MINIMUM_SHOULD_MATCH;
+    public static final float DEFAULT_BOOST_TERMS = 0;  // no boost terms
+    public static final boolean DEFAULT_INCLUDE = false;
+    public static final boolean DEFAULT_FAIL_ON_UNSUPPORTED_FIELDS = true;
+
+    // document inputs
+    private final List<String> fields;
+    private List<String> likeTexts = new ArrayList<>();
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java]: I think the add and set methods for those lists here are too complex. I think we should only have setters and the setters should make an unmodifiable list out of the values as well as copy them.

File: core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java [hunk 12]
```diff
@@ -85,9 +142,12 @@ public Item() {
          * @param type the type of the document
          * @param id and its id
          */
-        public Item(String index, @Nullable String type, String id) {
+        public Item(@Nullable String index, @Nullable String type, String id) {
             this.index = index;
             this.type = type;
+            if (id == null) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java]: can you move this up to fail before we assign

File: core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java [hunk 13]
```diff
@@ -98,10 +158,13 @@ public Item(String index, @Nullable String type, String id) {
          * @param type the type to be used for parsing the doc
          * @param doc the document specification
          */
-        public Item(String index, String type, XContentBuilder doc) {
+        public Item(@Nullable String index, @Nullable String type, XContentBuilder doc) {
             this.index = index;
             this.type = type;
-            this.doc(doc);
+            if (doc == null) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java]: can you move this up to fail before we assign

File: core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java [hunk 14]
```diff
@@ -517,6 +622,31 @@ public MoreLikeThisQueryBuilder maxWordLength(int maxWordLength) {
...
+     * <p>Any word in this set is considered "uninteresting" and ignored. Even if your Analyzer allows stopwords, you
+     * might want to tell the MoreLikeThis code to ignore them, as for the purposes of document similarity it seems
+     * reasonable to assume that "a stop word is never interesting".
+     */
+    public MoreLikeThisQueryBuilder stopWords(String... stopWords) {
+        this.stopWords = stopWords;
+        return this;
+    }
+
+    public MoreLikeThisQueryBuilder stopWords(List<String> stopWords) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java]: can we barf on null instead?

File: core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java [hunk 15]
```diff
@@ -679,4 +808,299 @@ private static void buildLikeField(XContentBuilder builder, String fieldName, Li
...
+
+        // possibly remove unsupported fields
+        removeUnsupportedFields(moreLikeFields, analyzerObj, failOnUnsupportedField);
+        if (moreLikeFields.isEmpty()) {
+            return null;
+        }
+        mltQuery.setMoreLikeFields(moreLikeFields.toArray(Strings.EMPTY_ARRAY));
+
+        // handle like texts
+        if (!likeTexts.isEmpty()) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java]: can you use `== false` instead of `!`

File: core/src/test/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilderTests.java [hunk 16]
```diff
@@ -0,0 +1,291 @@
...
+import java.util.Map;
+import java.util.stream.Stream;
+
+import static org.elasticsearch.common.xcontent.XContentFactory.jsonBuilder;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.is;
+
+public class MoreLikeThisQueryBuilderTests extends AbstractQueryTestCase<MoreLikeThisQueryBuilder> {
+
+    private static String[] randomFields;
```
Comment (s1monw) [inline core/src/test/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilderTests.java]: these don't need to be static

**Comment by s1monw** [comment]
I left some minor comments LGTM otherwise
